libraries/rush-sdk/src/proxy.ts

```diff
@@ -0,0 +1,288 @@
+    // SCENARIO 5: The rush-lib engine is loaded manually using rushSdkProxy.loadAsync().
+
+    if (!options) {
+      options = {};
+    }
+
+    const onNotifyEvent: SdkNotifyEventCallback | undefined = options.onNotifyEvent;
+    let progressBarPercent: number | undefined = undefined;
+
+    const abortSignal: AbortSignal = options.abortSignal ?? { aborted: false };
+
+    try {
+      const rushJsonSearchFolder: string = options.rushJsonSearchFolder ?? process.cwd();
+      const rushJsonPath: string | undefined = tryFindRushJsonLocation(rushJsonSearchFolder);
+      if (!rushJsonPath) {
+        throw new Error(
+          'Unable to find rush.json in the current folder or its parent folders.\n' +
+            'This tool is meant to be invoked from a working directory inside a Rush repository.'
+        );
+      }
+      const monorepoRoot: string = path.dirname(rushJsonPath);
+
+      const rushJson: JsonObject = JsonFile.load(rushJsonPath);
+      const { rushVersion } = rushJson;
+
+      const installRunNodeModuleFolder: string = path.join(
+        monorepoRoot,
+        `common/temp/install-run/@microsoft+rush@${rushVersion}`
+      );
+
+      try {
+        // First, try to load the version of "rush-lib" that was installed by install-run-rush.js
+        if (onNotifyEvent) {
+          onNotifyEvent({
+            state: 'running',
+            logMessage: {
+              messageType: 'info',
+              message: `Trying to load  ${RUSH_LIB_NAME} installed by install-run-rush`
+            },
+            progressBarPercent
+          });
+        }
+        sdkContext.rushLibModule = requireRushLibUnderFolderPath(installRunNodeModuleFolder);
+      } catch (e) {
+        let installAndRunRushStderrContent: string = '';
+        try {
+          const installAndRunRushJSPath: string = path.join(
+            monorepoRoot,
+            'common/scripts/install-run-rush.js'
+          );
+
+          if (onNotifyEvent) {
+            onNotifyEvent({
+              state: 'running',
+              logMessage: {
+                messageType: 'info',
+                message: 'The Rush engine has not been installed yet. Invoking install-run-rush.js...'
+              },
+              progressBarPercent
+            });
+          }
+
+          // Start the installation
+          progressBarPercent = 0;
+
+          const installAndRunRushProcess: SpawnSyncReturns<string> = Executable.spawnSync(
+            'node',
+            [installAndRunRushJSPath, '--help'],
+            {
+              stdio: 'pipe'
+            }
+          );
+
+          installAndRunRushStderrContent = installAndRunRushProcess.stderr;
+          if (installAndRunRushProcess.status !== 0) {
+            throw new Error(`The ${RUSH_LIB_NAME} package failed to install`);
+          }
+
+          if (!_checkForCancel(abortSignal, onNotifyEvent, progressBarPercent)) {
+            return;
+          }
```

Comment:
This pattern is a little strange. Aborting should result in the function rejecting with a "canceled" error, not returning `undefined`.

Reply:
Good feedback. I redesigned it to throw an `AbortError` exception